modules/table/src/main/java/org/apache/ignite/internal/table/distributed/disaster/ManualGroupUpdateRequest.java

```diff
@@ -79,31 +123,165 @@ public CompletableFuture<Void> handle(DisasterRecoveryManager disasterRecoveryMa
+     *
+     * @param tableDescriptor Table descriptor.
+     * @param zoneDescriptor Zone descriptor.
+     * @param dataNodes Current DZ data nodes.
+     * @param aliveNodesConsistentIds Set of alive nodes according to logical topology.
+     * @param revision Meta-storage revision to be associated with reassignment.
+     * @param metaStorageManager Meta-storage manager.
+     * @param localStatesMap Local partition states retrieved by {@link DisasterRecoveryManager#localPartitionStates(Set, Set, Set)}.
+     * @return A future that will be completed when reassignments data is written into a meta-storage, if that's required.
+     */
+    private CompletableFuture<?>[] forceAssignmentsUpdate(
+            CatalogTableDescriptor tableDescriptor,
+            CatalogZoneDescriptor zoneDescriptor,
+            Set<String> dataNodes,
+            Set<String> aliveNodesConsistentIds,
+            long revision,
+            MetaStorageManager metaStorageManager,
+            Map<TablePartitionId, LocalPartitionStateMessageByNode> localStatesMap
+    ) {
+        CompletableFuture<Map<Integer, Assignments>> tableAssignmentsFut = tableAssignments(
+                metaStorageManager,
+                tableDescriptor.id(),
+                partitionIds,
+                zoneDescriptor.partitions()
+        );
+
+        Set<String> aliveDataNodes = CollectionUtils.intersect(dataNodes, aliveNodesConsistentIds);
+
+        int[] partitionIdsArray = partitionIds.isEmpty()
+                ? IntStream.range(0, zoneDescriptor.partitions()).toArray()
+                : partitionIds.stream().mapToInt(Integer::intValue).toArray();
+
+        CompletableFuture<?>[] futures = new CompletableFuture[partitionIdsArray.length];
+
+        for (int partitionId = 0; partitionId < partitionIdsArray.length; partitionId++) {
+            TablePartitionId replicaGrpId = new TablePartitionId(tableDescriptor.id(), partitionIdsArray[partitionId]);
+
+            futures[partitionId] = tableAssignmentsFut.thenCompose(tableAssignments ->
+                    tableAssignments.isEmpty() ? nullCompletedFuture() : manualPartitionUpdate(
+                            replicaGrpId,
+                            aliveDataNodes,
+                            aliveNodesConsistentIds,
+                            zoneDescriptor.replicas(),
+                            revision,
+                            metaStorageManager,
+                            tableAssignments.get(replicaGrpId.partitionId()).nodes(),
+                            localStatesMap.get(replicaGrpId)
+                    )).thenAccept(res -> {
+                        DisasterRecoveryManager.LOG.info(
+                                "Partition {} returned {} status on reset attempt", replicaGrpId, UpdateStatus.valueOf(res)
+                        );
+                    }
+            );
+        }
+
+        return futures;
+    }
+
+    private static CompletableFuture<Integer> manualPartitionUpdate(
```

Comment:
The method is more difficult to read because it contains many lines of code, I would divide it into several methods, but at your discretion.